Sources/LanguageServerProtocol/Notifications/SourceKitFileStatusNotification.swift

```diff
@@ -0,0 +1,48 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// **LSP Extension** Notification from the server about a per-file status update for
+/// the client.
+///
+/// Requires a private SourceKit extension.
+public struct SourceKitFileStatusNotification: TextDocumentNotification, Hashable {
+  public static let method: String = "textDocument/sourcekit.filestatus"
+
+  /// The `textDocument` this status notification is for.
+  public var textDocument: TextDocumentIdentifier
+
+  /// File state used to stylize the status (e.g. with an icon).
+  public var state: SourceKitFileState
+
+  /// Severity of the status. Used to stylize the status (e.g. with a color or icon).
+  public var severity: DiagnosticSeverity
```

Comment:
Should this be optional and set to `nil` for the "all-OK state"?

I like the general idea of just classifying as "warning" when it is known that there may be issues. For "error" I'd probably only use it for really catastrophic events like the build service crashing or something.

Reply:
Yeah, I was thinking this would only be used for warnings and errors, I could make it a separate type. Example cases I was thinking:

* Warning: file using fallback build settings or needs prebuilding
* Error: Internal error or invalid build set up (e.g. error in Package.swift, error in BUILD file)

I separated this from the working vs. ready status to denote loading when in the error / warning state (e.g. if reloading Package.swift files). In VS Code, you can use a separate color to handle the severity (or an icon), which makes the warning/error states more pronounced. Let me know which you'd prefer.